x-pack/platform/plugins/shared/onechat/server/services/tools/tool_provider.ts

```diff
@@ -47,67 +48,36 @@ export interface ToolUpdateParams<TConfig extends object = {}> {
 export type ToolTypeCreateParams<TConfig extends object = {}> = ToolCreateParams<TConfig>;
 export type ToolTypeUpdateParams<TConfig extends object = {}> = ToolUpdateParams<TConfig>;
 
-export interface ToolSourceGetClientArgs {
-  request: KibanaRequest;
-  space: string;
-}
-
-/**
- * Defines a provider for a given tool type
- */
-export type ToolSource<T extends ToolType = ToolType> = {
-  /**
-   * Unique identifier for this source
-   */
+export interface ReadonlyToolProvider {
   id: string;
-  /**
-   * List of tool types provided by this source
-   */
-  toolTypes: T[];
-} & (
-  | {
-      readonly?: false | undefined;
-      getClient(opts: ToolSourceGetClientArgs): MaybePromise<ToolTypeClient>;
-    }
-  | {
-      readonly: true;
-      getClient(opts: ToolSourceGetClientArgs): MaybePromise<ReadonlyToolTypeClient>;
-    }
-);
-
-export interface ToolTypeClient<TConfig extends object = {}>
-  extends ReadonlyToolTypeClient<TConfig> {
-  /**
-   * Create a new tool
-   */
-  create(params: ToolTypeCreateParams<TConfig>): MaybePromise<InternalToolDefinition<TConfig>>;
-  /**
-   * Update an existing tool
-   */
-  update(
-    toolId: string,
-    update: ToolTypeUpdateParams<TConfig>
-  ): MaybePromise<InternalToolDefinition<TConfig>>;
+  readonly: true;
+  has(toolId: string): MaybePromise<boolean>;
+  get(toolId: string): MaybePromise<InternalToolDefinition>;
+  list(): MaybePromise<Array<InternalToolDefinition>>;
+}
 
-  /**
-   * Delete a tool
-   */
+export interface WritableToolProvider extends Omit<ReadonlyToolProvider, 'readonly'> {
+  readonly: false;
+  create(params: ToolTypeCreateParams): MaybePromise<InternalToolDefinition>;
+  update(toolId: string, update: ToolTypeUpdateParams): MaybePromise<InternalToolDefinition>;
   delete(toolId: string): MaybePromise<boolean>;
```

Comment:
Re-using the pattern of write/read provider that is already being used for agents, as I was fairly happy with it, and it's better than what I built previously for tools